### ISSUE TITLE
Using BIO_set_mem_buf would leak in 1.1

### ISSUE THREAD

**Issue by veselov** [openssl/openssl#8025]
Maybe I'm missing something here, I have googled around, and found practically nothing about this, which is suspicious... 

I have this code that works fine in 1.0:
```
BUF_MEM p7Buf = { .data = p7Data , .length = p7Len, .max = p7Len };
BIO * bioMem = BIO_new(BIO_s_mem());
BIO_set_mem_buf(bioMem, &p7Buf, BIO_NOCLOSE);
// do stuff with bioMem
BIO_free(bioMem);
```

In 1.1 (1.1.0i) this code starts leaking. This is because `mem_method->mem_free` only agrees to free anything it created if `bio->shutdown` is `!0`. `bio->shutdown` is set to `0` when `BIO_set_mem_buf` is called with `BIO_NOCLOSE`, so when I call `BIO_free`, none of the stuff in `bio->ptr` is released.

I don't find this particularly reasonable, since nobody from outside the mem knows what's in `bio->ptr`, and wouldn't really know how to release any of it. This also would seem to contradict the documentation of `BIO_S_MEM(3)`, which says that the above pattern is good for having your buffer being left alone, but I'd think the assumption that all other stuff will be properly cleaned up.

I understand that the migration path here is to allocate the buffer, instead of creating it statically, set the buffer with `BIO_CLOSE` (so that `bio->shutdown` remains `!0`), but set the `BIO_FLAGS_MEM_RDONLY` so that the buffer data doesn't get freed.

**Comment by davidben** [comment]
Yeah, this does look like a regression from 9fe9d0461ea4bcc42cd75a30f013fa61b5407b93. The `BIO_NOCLOSE` passed to `BIO_set_mem_buf` ought to mean that the caller is not passing ownership to the `BUF_MEM`, but that doesn't imply anything about internal details of how the `BIO` stores the pointer. It turns out the `BIO` now allocates a separate wrapper structure, but that's not the caller's business.

**Comment by cminyard** [comment]
I ran into the same thing, I've attached a pull request that fixes the issue.  The following program reproduces, but there's no real way to do a real test for it without using something like valgrind.

Pull request is #8037 , I've run it through the standard tests.

#include <stdio.h>
#include <string.h>
#include <openssl/buffer.h>
#include <openssl/bio.h>

int main(int argc, char *argv[])
{
	BIO *tbio;
	BUF_MEM tbufmem;
	const char *str = "BIO test\n";
	char buf[100];

	tbio = BIO_new(BIO_s_mem());
	tbufmem.length = strlen(str) + 1;
	tbufmem.data = (char *) str;
	tbufmem.max = tbufmem.length;
	BIO_set_mem_buf(tbio, &tbufmem, BIO_NOCLOSE);
	BIO_set_flags(tbio, BIO_FLAGS_MEM_RDONLY);

	if (BIO_read(tbio, buf, sizeof(buf)) <= 0) {
		fprintf(stderr, "Error reading BIO\n");
		return 1;
	}
	printf("%s", buf);
	BIO_free(tbio);
	return 0;
}

**Comment by t8m** [comment]
This should be fully resolved in 1.1.1c.
